README

abstract:
  wdfs is a webdav filesystem with special features for accessing subversion
  repositories. it is based on fuse v2.5+ and neon v0.24.7+.


author of wdfs:
  (c) 2005 - 2007 jens m. noedler, noedler@web.de


license:
  wdfs is distributed under the terms of the gnu general public license.
  see file COPYING for details.


home and svn repository of wdfs:
  http://noedler.de/projekte/wdfs/
  http[s]://svn.projectai.org/svn/projects_private/wdfs/


features:
 - generic webdav filesystem
 - http redirect support
 - https support
 - file locking support (different modes)
 - access to all revisions of a webdav exported subversion repository
 - versioning filesystem for autoversioning enabled subversion repositories
   (see section "wdfs, subversion and apache" in this document)


dependencies:
 - operating systems: linux (kernel 2.4 or 2.6), freebsd (6.x or 7.x), mac os x
   hint for linux user: using a kernel 2.6.15 or later is recommended.
 - fuse (filesystem in userspace) v2.5 or later (http://fuse.sourceforge.net/)
 - neon webdav library v0.24.7 or later (http://www.webdav.org/neon/)


installation:
  $ tar xfz wdfs-1.x.x.tar.gz
  $ cd wdfs-1.x.x
  $ ./configure && make
  $ su -c "make install"


usage:
  $ mkdir ~/mountpoint
  $ wdfs http[s]://server[/directory/] ~/mountpoint [options]
  $ wdfs -h      # prints wdfs and fuse options


mounting wdfs using fstab:
  example entry for the /etc/fstab file:
  wdfs#http://server  /mnt/wdfs  fuse  users,locking=simple  0  0

  hint: due to a bug in fuse 2.6.3 and earlier it's not possible to add
  the option "username=abc" to the option list. this should be fixed with 
  fuse 2.6.4.


wdfs, subversion and apache:
  wdfs may be used to access subversion repositories that are exported as
  webdav using the apache webserver. you need to load the required apache 
  modules for webdav and for subversion and add a location directive like
  the following.

  <Location "/svn">
    DAV svn
    SVNPath /home/jens/repositories/wdfs
    SVNAutoversioning on
    AuthType None
  </Location>


version number conventions:
 - the major number will never change.
 - the minor number changes if new features are introduced or 
   if parts of the api are changed.
   warning: this may include backward-incompatible changes!
 - the last number changes on bug fixes and little non-api changes.
   e.g. v1.1.0 introduced new features from it's predecessor v1.0.1


limitations of this implementation:
 - wdfs is not (yet) multithread safe. it always uses fuse single thread mode.
 - read/write of big files (~ 50 MB+) is slow, because the _complete_ file
   will be GET/PUT on a request.
 - svn mode: only up to 2^31-1 revisions can be accessed,
   because "latest_revision" is an integer variable.
 - svn mode: if a subdirectory of a subversion repository is mounted, the
   "svn_basedir" shows the content of the repository root and not only the
   content of the mounted subdir. [caused by svn_get_remotepath()]
 - http redirect support:
    * there is no redirect support for the mountpoint. please mount the new
      location, that will be printed as an error message.
    * the redirect support is limited to the webdav server, that is mounted.
      redirects to another server are not supported.


limitations of wdfs, caused by other limitations:
 - wdfs supports only utf-8 encoded file and directory names for svn access.
   use utf8 capable applications or pure ascii (no iso8859!) encoded names.
   (limitation caused by subversion, which stores all data utf-8 encoded)
 - wdfs does not support setting file attributes with utime() or utimes().
   (e.g. used by "touch file -m -t 200601010000"). see wdfs_setattr() for
   details. (limitation caused by webdav specification rfc 2518)


source code conventions:
 - variable name "remotepath" is used for WebDAV access (/remote/dir/file)
 - variable name "localpath" is used for local access (/file)
 - variable name "ret" is used for the return values of methods
 - methods and structs starting with "ne_" are part of the neon library
 - error output:
    - "## " prefix for warnings and error messages
 - debug output:
    - ">> " prefix for fuse callback methods
    - "** " prefix for cache related messages (no errors)
    - "++ " prefix for locking related messages (no errors)
 - the code is formated with tabs (tab width 4 characters)
 - every line should not be longer than 80 characters


acknowledgements:
  helmut neukirchen for supervising my bachelor thesis and the development
  of wdfs. miklos szeredi for developing fuse and being a great maintainer.
  the members of the fuse-devel mailing list for helping me and everybody 
  who is contributing to wdfs. thanks!





ChangeLog
2007-02-05  noedler

	- wdfs, 1.4.1
	- updated README
	- added redirect handling for rename() and unlink(). thanks joseph!

2007-01-25  noedler

	compile fix for mac os x: define strndup() for apple computers

2007-01-24  noedler

	fixed mounting a webdav share that redirects (e.g.) from '/foo' to '/foo/'. thanks to joseph dunn for the patch.

2007-01-19  noedler

	wdfs, 1.4.0

2007-01-14  noedler

	enable http redirect support by default and add option "no_redirect" to disable it

2007-01-14  noedler

	coding style...

2007-01-13  noedler

	- updated copyright to 2007
	- added dummy impl of wdfs_init()
	- updated README with an example how to mount wdfs thru fstab

2007-01-13  noedler

	updated README file

2006-11-13  noedler

	added a dummy implementation for statfs()

2006-11-09  noedler

	fixed a bug in the assert statements, thanks sven bachmann for pointing it out!

2006-10-27  noedler

	fixed creating the ChangeLog

2006-10-27  noedler

	Initial commit for the public SVN

2006-10-17  jens m. noedler  <noedler@web.de>

	- introduced unify_path() which does the escaping/unescaping
	  the remotepath in one place
	- added checks for the return value of remove_ending_slash()
	- new: remotepath_basedir may contain special chars
	- some random coding style improvements

2006-10-17  jens m. noedler  <noedler@web.de>

	improvement: return -EPERM instead of -ENOENT if deleting fails
	with "403 Forbidden". patch from joseph.dunn@colorado.edu

2006-10-16  jens m. noedler  <noedler@web.de>

	just some random coding style improvements

2006-10-14  jens m. noedler  <noedler@web.de>

	- fixed parsing of "-o accept_sslcert" parameter
	- added check of "-o locking_timeout" parameter

2006-10-14  jens m. noedler  <noedler@web.de>

	converted most printf() to fprintf(stderr, ...)

2006-10-14  jens m. noedler  <noedler@web.de>

	also allow to enter authentication information (username and password)
	via the terminal, based on a patch from joseph.dunn@colorado.edu. thanks!

2006-10-13  jens m. noedler  <noedler@web.de>

	- new command line parsing using the fuse option functions
	  thanks to yuri vasilevski for the patch!
	- wdfs require due to this change fuse 2.5 or later

2006-08-22  jens m. noedler  <noedler@web.de>

	- replaced "DT_DIR << 12" with "S_IFDIR | 0777" for "." and ".." dentries
	- removed useless includes of time.h

2006-08-21  jens m. noedler  <noedler@web.de>

	bugfix: set . and .. to type DT_DIR
	thanks for the patch from mathieu boespflug

2006-08-01  jens m. noedler  <noedler@web.de>

	wdfs, 1.3.1

2006-08-01  jens m. noedler  <noedler@web.de>

	added these statement to permit debian to add wdfs to it's archive

	 *  This program is released under the GPL with the additional exemption
	 *  that compiling, linking and/or using OpenSSL is allowed.

	details: http://www.gnome.org/~markmc/openssl-and-the-gpl.html

2006-07-26  jens m. noedler  <noedler@web.de>

	wdfs, 1.3.0

2006-07-26  jens m. noedler  <noedler@web.de>

	implemented dummy wdfs_chmod() method

2006-07-26  jens m. noedler  <noedler@web.de>

	fixed some escape/unescape problems

2006-07-26  jens m. noedler  <noedler@web.de>

	fix: invalid read in remove_ending_slash()

2006-07-25  jens m. noedler  <noedler@web.de>

	simplified #definition of NEON_VERSION

2006-07-25  jens m. noedler  <noedler@web.de>

	- do not allow lock_timeout to be set to 0 seconds
	- fixed a bug regarding too late unescaping of the remotepath,
	  what may lead to recursive adding of the directory to itself.

2006-07-25  jens m. noedler  <noedler@web.de>

	memory management: replaced wdfs_free() through FREE() and free_chars()

2006-07-25  jens m. noedler  <noedler@web.de>

2006-07-24  jens m. noedler  <noedler@web.de>

	wdfs is now compatible to neon 0.26, thanks again to suse

2006-07-21  jens m. noedler  <noedler@web.de>

	type updates and conversion of error_parameter_parsing from int to bool_t

2006-07-21  jens m. noedler  <noedler@web.de>

	correct indentation and fix for "the last parameter parsing bug" ;-)

2006-07-20  jens m. noedler  <noedler@web.de>

	typo updates

2006-07-20  jens m. noedler  <noedler@web.de>

	conversion from NE_FREE(...) to wdfs_free(...)

2006-07-20  jens m. noedler  <noedler@web.de>

	enhanced argument handling, also based on suse patches

2006-07-20  jens m. noedler  <noedler@web.de>

	reset arguments to avoid storing sensitive data in the process table
	based on a patch from max@suse.de

2006-07-20  jens m. noedler  <noedler@web.de>

	fix for 64 bit platforms due to missing braces
	patch from max@suse.de, thanks!

2006-06-27  jens m. noedler  <noedler@web.de>

	added a comment about freebsd compat in the README

2006-06-27  jens m. noedler  <noedler@web.de>

	wdfs, 1.2.1
	just a build fix for freebsd

2006-06-26  jens m. noedler  <noedler@web.de>

	compile fix for freebsd

2006-04-10  jens m. noedler  <noedler@web.de>

	wdfs, 1.2.0
	updated the README

2006-04-10  jens m. noedler  <noedler@web.de>

	finally fixed the last ftruncate() bug
	thanks to miklos! :-)

2006-04-06  jens m. noedler  <noedler@web.de>

	ftruncate() done.
	but there is still a bug with ftruncate() and O_APPEND

2006-03-12  jens m. noedler  <noedler@web.de>

	made redirect support configurable via parameter "-r"
	("-r" is no longer used by fuse, I think since v2.5)

2006-03-12  jens m. noedler  <noedler@web.de>

	smoother redirect support using
	static int handle_redirect(char **remotepath);

2006-03-12  jens m. noedler  <noedler@web.de>

	stub for wdfs_ftruncate

2006-03-12  jens m. noedler  <noedler@web.de>

	some changes for FUSE_USE_VERSION 25
	and compat with fuse 2.3 and 2.4

2006-03-11  jens m. noedler  <noedler@web.de>

	first stage of redirect support
	first steps to go to fuse interface v25

2006-01-05  jens m. noedler  <noedler@web.de>

	wdfs, 1.1.1
	added some comments about the fixed bug

2006-01-04  jens m. noedler  <noedler@web.de>

	1.1.1-pre1

2006-01-04  jens m. noedler  <noedler@web.de>


src/svn.c
/* 
 *  this file is part of wdfs --> http://noedler.de/projekte/wdfs/
 *
 *  wdfs is a webdav filesystem with special features for accessing subversion
 *  repositories. it is based on fuse v2.3+ and neon v0.24.7+.
 * 
 *  copyright (c) 2005 - 2006 jens m. noedler, noedler@web.de
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 *  This program is released under the GPL with the additional exemption
 *  that compiling, linking and/or using OpenSSL is allowed.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <assert.h>
#include <unistd.h>
#include <glib.h>
#include <ne_props.h>

#include "wdfs-main.h"
#include "webdav.h"
#include "svn.h"


/* wdfs has some special subversion (svn) related features. if "svn_mode" is
 * set to "true", wdfs will provide transparent access to all svn revisions via
 * a virtual directory, that will be added below the fuse mountpoint. the name
 * of the virtual directory is set to the value of the variable "svn_basedir".
 * 
 * below the svn_basedir there will be a dirctory for each revision, that
 * contains all files, that belong to the specific revision. because there 
 * can be a lot of revision, the directories will be clustered into chunks.
 * 
 * the directory chunks are called "level 1" and the ones, which contains the
 * specific revision's data are called "level 2":
 *  
 * level 2 -----------------\
 * level 1 ---------\       |
 *                  |       |
 * ...              |       |
 * /svn_basedir/5000-5199/5000/
 * /svn_basedir/5000-5199/5001/
 * ...
 * /svn_basedir/5000-5199/5119/
 * /svn_basedir/5200-5399/5200/
 * /svn_basedir/5200-5399/5201/
 * ...
 * 
 */


/* svn_basedir specifies the name of the virtual directory, that allows
 * accessing all svn revisions. it must start with '/' and end without '/'. 
 * it may be edited here. */
const char *svn_basedir = "/0-all-revisions";

/* svn_repository_root is the path to the root of the svn repository. this is
 * needed, if a subdir of the repository is mounted to allow access to the old
 * revisions, because the path to access them always starts at the root of a 
 * repository. */
char *svn_repository_root = NULL;

/* controls how many directories are put in a single level 2 chunk. editable. */
static const int svn_revisions_per_level2_directory = 200;

/* svn_mode enables transparent access to all svn revisions in a repository
 * via a virtual directory. "true" enables svn_mode, "false" disables svn_mode.
 * don't edit here! it can be changed via parameter "-S" passed to wdfs. */
bool_t svn_mode = false;


/* webdav properties used to get the latest svn revision */
static const ne_propname property_checked_in[] = {
	{ "DAV:", "checked-in"},
	{ NULL } /* MUST be NULL terminated! */
};

static const ne_propname property_vcc[] = {
	{ "DAV:", "version-controlled-configuration"},
	{ NULL } /* MUST be NULL terminated! */
};



/* +++++++ local static methods +++++++ */

/* author jens, 03.08.2005 14:39:23, location: goettingen */


/* this function extracts the revision number from a string */
static void svn_get_latest_revision_callback(
#if NEON_VERSION >= 26
	void *userdata, const ne_uri* href_uri, const ne_prop_result_set *results)
#else
	void *userdata, const char *href, const ne_prop_result_set *results)
#endif
{
	assert(userdata && results);

	int *latest_revision = (int*)userdata;
	char *latest_revision_string = NULL;
	const char delimiters[] = "/";
	char *token = NULL;
	char *result = (char*)ne_propset_value(results, &property_checked_in[0]);
	/* result now looks like this "<href>/svn/dir1/dirx/!svn/bln/1234</href>".
	 * extracting the number in a bit tricky. split the string at each '/'
	 * and check if it's "!svn" followed by "bln" followed by the number. */
	do {
		token = strsep(&result, delimiters);
		if (token != NULL && !strcmp("!svn", token)) {
			/* next token should be bln */
			token = strsep(&result, delimiters);
			if (token != NULL && !strcmp("bln", token)) {
				/* the next token is the svn revision */
				token = strsep(&result, delimiters);
				if (token != NULL) {
					/* remove the last char; e.g. "1234<" to "1234" */
					latest_revision_string = strndup(token, strlen(token)-1);
				}
			}
		}
	} while (token != NULL );

	if (debug_mode == true)
		printf(">> SVN latest revision _string_: %s\n", latest_revision_string);

	/* string to integer conversion */
	*latest_revision = atoi(latest_revision_string);

	FREE(latest_revision_string);
}


/* returns -1 on error or the latest svn revision on success */
static int svn_get_latest_revision()
{
	int latest_revision;
	char *uri = ne_concat(svn_repository_root, "!svn/vcc/default", NULL);
	ne_propfind_handler *pfh = ne_propfind_create(session, uri, NE_DEPTH_ZERO);
	int ret = ne_propfind_named(pfh, property_checked_in,
					&svn_get_latest_revision_callback, &latest_revision);
	ne_propfind_destroy(pfh);
	FREE(uri);
	if (ret != NE_OK) {
		printf("## ne_propfind_named() error\n");
		return -1;
	}
	return latest_revision;
}


/* return the number of '/' found in a given string */
static int svn_directory_depth(const char *in)
{
	assert(in);

	int i, ret = 0;
	int len = strlen(in);
	for (i = 0; i < len; i++) {
		if (in[i] == '/')
			ret++;
	}
	return ret;
}


/* like strlen(), but for integers */
static int get_integer_length(int in)
{
	assert(in);

	int length = 1;
	while (in >= 10) {
		in = in / 10;
		length++;
	}
	return length;
}


/* callback from svn_set_repository_root() to do the dirty string parsing. */
static void svn_set_repository_root_callback(
#if NEON_VERSION >= 26
	void *userdata, const ne_uri* href_uri, const ne_prop_result_set *results)
#else
	void *userdata, const char *href, const ne_prop_result_set *results)
#endif
{
	assert(results);

	/* the string looks like "<href>/.../.../!svn/vcc/default</href>" */
	char *result = (char*)ne_propset_value(results, &property_vcc[0]);
	/* use a GString struct for comfortable string erasing */
	GString *g_result = g_string_new(result);
	/* remove starting tag ("<href>") */
	g_string_erase(g_result, 0, 6);
	/* remove trailing stuff ("!svn/vcc/default</href>"),
	 * so that only "/.../.../" -- the repository root path -- remains. */
	g_string_erase(g_result, g_result->len - 23, 23);
	svn_repository_root = strdup(g_result->str);
	g_string_free(g_result, TRUE);
}


/* +++++++ exported non-static methods +++++++ */

/* set the root path of the mounted repository. this path might differ from the
 * mounted uri if a subdir of the repository is mounted. return 0 on success or
 * -1 on error. */
int svn_set_repository_root() {
	/* if remotepath_basedir is empty set svn_repository_root to "/" and quit */
	if (!strcmp(remotepath_basedir, "")) {
		svn_repository_root = strdup("/");
		return 0;
	}

	ne_propfind_handler *pfh = 
		ne_propfind_create(session, remotepath_basedir, NE_DEPTH_ZERO);
	int ret = ne_propfind_named(pfh, property_vcc, 
		&svn_set_repository_root_callback, NULL);
	ne_propfind_destroy(pfh);
	if (ret != NE_OK) {
		printf("## ne_propfind_named() error\n");
		return -1;
	}
	return 0;
}

void svn_free_repository_root() {
	FREE(svn_repository_root);
}


/* converts a localpath to a remotepath to access old revision 
 * IN:              /svn_basedir/x-y/1234/directory/file.txt
 * OUT: /svn_repository_root/!svn/bc/1234/directory/file.txt or NULL on error */
char* svn_get_remotepath(const char *localpath)
{
	assert(localpath);
	/* use a GString struct for comfortable string erasing */
	GString *g_localpath = g_string_new(localpath);
	/* remove the svn_basedir and the next char from the start of the string */
	g_string_erase(g_localpath, 0, strlen(svn_basedir) + 1);
	/* save the string starting with the next '/', because it's the relative
	 * path including the revision. e.g. "/1234/directory/file.txt" */
	char *path = strchr(g_localpath->str, '/');
	/* concat the svn uri string, that allows to access this revision */
	char *remotepath = ne_concat(svn_repository_root, "!svn/bc", path, NULL);
	g_string_free(g_localpath, TRUE);
	if (remotepath == NULL)
		return NULL;
	/* finally escape the string */
	char *remotepath2 = ne_path_escape(remotepath);
	FREE(remotepath);
	if (remotepath2 == NULL)
		return NULL;
	return remotepath2;
}


/* return a static stat, used for svn_basedir and level 1 directories */
struct stat svn_get_static_dir_stat()
{
	struct stat stat;
	stat.st_mode	= S_IFDIR | 0777; 
	stat.st_size	= 4096;
	stat.st_nlink	= 1;
	stat.st_atime	= stat.st_mtime = stat.st_ctime = time(NULL);
	stat.st_blocks	= (stat.st_size+511)/512;
	stat.st_mode	&= ~umask(0);
	stat.st_uid		= getuid();
	stat.st_gid		= getgid();
	return stat;
}


/* gets the latest revision number from subversion and adds the level1-dirs */
void svn_add_level1_directories(struct dir_item *item_data)
{
	assert(item_data);

	int latest_revision = svn_get_latest_revision();
	if (latest_revision >= 0) {

		GString *int2string = 
			g_string_new_len("", get_integer_length(latest_revision));

		int i, rest = latest_revision % svn_revisions_per_level2_directory;
		for (i = 0; i <= latest_revision; i++) {
			if (i % svn_revisions_per_level2_directory == 0) {
				GString *directory_name = g_string_new("");
				sprintf(int2string->str, "%d", i);
				g_string_append(directory_name, int2string->str);
				g_string_append_c(directory_name, '-'); 
				if (i + svn_revisions_per_level2_directory > latest_revision)
					sprintf(int2string->str, "%d", i + rest );
				else
					sprintf(int2string->str, "%d", 
						i + (svn_revisions_per_level2_directory-1) );
				g_string_append(directory_name, int2string->str);
				item_data->filler(item_data->buf, directory_name->str, NULL, 0);
				g_string_free(directory_name, TRUE);
			}
		}
		g_string_free(int2string, TRUE);
	} else {
		printf("## Error: Could not get latest revision from SVN.\n");
	}
}


/* sets the stat for a level1 directory and returns 0 (success) or 1 (error). */
int svn_get_level1_stat(struct stat *stat, const char *localpath) {
	assert(stat && localpath);

	if (svn_directory_depth(localpath) == 2) {
		*stat = svn_get_static_dir_stat();
		return 0;
	}
	return 1;
}


/* IN:  /svn_basedir/x-y/        (as string, x and y are revision numbers)
 * OUT: /svn_basedir/x-y/x/      (as level2-dentry, added with filler method)
 *      /svn_basedir/x-y/x+1/    (as level2-dentry, added with filler method)
 *      /svn_basedir/x-y/../     (as level2-dentry, added with filler method)
 *      /svn_basedir/x-y/y-1/    (as level2-dentry, added with filler method)
 *      /svn_basedir/x-y/y/      (as level2-dentry, added with filler method)
 * returns 0 on success (added level2 directories) and 1 otherwise. */
int svn_add_level2_directories(
	struct dir_item *item_data, const char *localpath)
{
	assert(item_data && localpath);

	if (svn_directory_depth(localpath) == 2) {
		/* use a GString object for comfortable string erasing */
		GString *g_localpath = g_string_new(localpath);
		/* remove the svn_basedir and the next char from the start */
		g_string_erase(g_localpath, 0, strlen(svn_basedir) + 1);

		char delimiters[] = "-";
		char *token;
		/* the variable x_y contains the string "x-y", x and y are numbers */
		char *x_y = strdup(g_localpath->str);
		g_string_free(g_localpath, TRUE);

		char *pointer = x_y;
		token = strsep(&x_y, delimiters);
		int x = atoi(token);
		token = strsep(&x_y, delimiters);
		int y = atoi(token);
		FREE(pointer);

		GString *int2string = g_string_new_len("", get_integer_length(y));
		int i;
		/* insert all level2-dentries from x to y */
		for (i = x; i <= y; i++) {
			/* convert int to string by printing it into a string object */
			sprintf(int2string->str, "%d", i);
			item_data->filler(item_data->buf, int2string->str, NULL, 0);
		}
		g_string_free(int2string, TRUE);
		return 0;
	}
	return 1;
}



src/cache.c
/* 
 *  this file is part of wdfs --> http://noedler.de/projekte/wdfs/
 *
 *  wdfs is a webdav filesystem with special features for accessing subversion
 *  repositories. it is based on fuse v2.3+ and neon v0.24.7+.
 * 
 *  copyright (c) 2005 - 2006 jens m. noedler, noedler@web.de
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 *  This program is released under the GPL with the additional exemption
 *  that compiling, linking and/or using OpenSSL is allowed.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <assert.h>
#include <glib.h>
#include <pthread.h>
#include <unistd.h>

#include "wdfs-main.h"
#include "cache.h"

/* this cache is designed to buffer the file's attributes (struct stat) locally
 * instead of sending a new request to the webdav server. this leads into a
 * better responsiveness of the filesystem.
 * every file's attributes is stored in a 'struct cache_item' that contains a
 * 'struct stat' and a 'time_t timeout' field. the timeout field is used to 
 * purge the cache_item, if it is too old. how long a cache_item is stored is
 * configured at CACHE_ITEM_TIMEOUT (in seconds).
 * the cache_items are stored in a hash as a hash value and the corresponding
 * remotepath (uri) as the hash key. because the remotepath (uri) is unique,
 * a hash is a appropriate data structure and accessing it is fast.
 * a 2nd thread runs every CACHE_ITEM_TIMEOUT seconds in the background and
 * removed timed out cache_items. 
 */


/* lifetime of a cache item in seconds. this value can be edit here. */
#define CACHE_ITEM_TIMEOUT 20

/* initalize this mutex, which is used to prevent data 
 * inconsistencies due to race conditions. */
pthread_mutex_t cache_mutex = PTHREAD_MUTEX_INITIALIZER;

/* every created thread needs an id. this is the cache control thread's id. */
pthread_t cache_control_thread_id;

/* hash object to store the cache items */
static GHashTable *cache;


struct cache_item {
	struct stat stat;	/* 96 bytes (i386) */
	time_t timeout;		/*  4 bytes (i386) */
};


/* +++++++ local static methods +++++++ */
/* author jens, 31.07.2005 18:44:28, location: heli at heinemanns */


/* returns 0 if the item is _not_ timed out, returns 1 if it _is_ timed out. */
static int cache_item_timed_out(const int timeout)
{
	if (timeout - time(NULL) <= 0)
		return 1;
	else 
		return 0;
}


/* callback method for g_hash_table_foreach_remove() called in cache_control_
 * thread(). this method check if a cache item has reached it's timeout and 
 * then removes it from the cache.  */
static int cache_control_thread_callback(void *key, void *value, void *userdata)
{
	struct cache_item *item = (struct cache_item *)value;
	if (cache_item_timed_out(item->timeout)) {
		if (debug_mode == true) {
			printf("** cache control thread: ");
			printf("item has timed out and is removed '%s'\n", (char*)key);
		}
		/* remove this item */
		return 1;
	}
	/* don't remove this item */
	return 0;
}


/* +++++++ exported non-static methods +++++++ */


/* this thread runs until it is canceled by the main thread and
 * removes every CACHE_ITEM_TIMEOUT seconds timed out cache items. */
static void* cache_control_thread(void *unused)
{
	pthread_setcancelstate(PTHREAD_CANCEL_ENABLE, NULL);
	/* thread exits if a cancel signal is recieved and
	 * cansel state is PTHREAD_CANCEL_ENABLE */
	pthread_setcanceltype(PTHREAD_CANCEL_ASYNCHRONOUS, NULL);
	while (1) {
		sleep(CACHE_ITEM_TIMEOUT);
		/* do not allow cancling this thread while doing it's work */
		pthread_setcancelstate(PTHREAD_CANCEL_DISABLE, NULL);
		/* to avoid conflict with cache_delete_item() lock */
		pthread_mutex_lock(&cache_mutex);
		/* check each cache item, if it is timed out and remove it */
		g_hash_table_foreach_remove(cache, &cache_control_thread_callback, NULL);
		pthread_mutex_unlock(&cache_mutex);
		/* now this thread might be cancel, because it is idle */
		pthread_setcancelstate(PTHREAD_CANCEL_ENABLE, NULL);
	}
	return NULL;
}


/* initializes the cache's hash table and start a 2nd thread, that removed 
 * timed out item from the cache periodically. */
void cache_initialize()
{
	cache = g_hash_table_new_full(g_str_hash, g_str_equal, g_free, g_free);
	assert(cache);

	/* setup a thread, that removes timed out cache items in the background */
	pthread_create(&cache_control_thread_id, NULL, &cache_control_thread, NULL);
}


/* detroys the cache if it's no longer needed. joins the 2nd thread and kills
 * the hash table. */
void cache_destroy()
{
	if (debug_mode == true)
		printf("** destroying %d cache items\n", g_hash_table_size(cache));
	/* exit cache control thread */
	pthread_cancel(cache_control_thread_id);
	pthread_join(cache_control_thread_id, NULL);

	/* to avoid conflict with cache_control_thread() lock */
	pthread_mutex_lock(&cache_mutex);
	g_hash_table_destroy(cache);
	pthread_mutex_unlock(&cache_mutex);
}


/* adds a new item to the cache and sets the items timeout. */
void cache_add_item(struct stat *stat, const char *remotepath)
{
	assert(remotepath && stat);

	/* generalize string by removing the ending slash */
	char *remotepath2 = remove_ending_slash(remotepath);

	/* get the new cache item and set it's values */
	struct cache_item *item = g_new0(struct cache_item, 1);
	item->stat = *stat;
	item->timeout = time(NULL) + CACHE_ITEM_TIMEOUT;

	/* to avoid conflict with cache_control_thread() lock */
	pthread_mutex_lock(&cache_mutex);
	g_hash_table_insert(cache, strdup(remotepath2), item);
	pthread_mutex_unlock(&cache_mutex);

	if (debug_mode == true)
		printf("** added cache item for '%s'\n", remotepath2);
	FREE(remotepath2);
}


/* deletes a cache item from the cache. */
void cache_delete_item(const char *remotepath)
{
	assert(remotepath);

	/* generalize string by removing the ending slash */
	char *remotepath2 = remove_ending_slash(remotepath);

	/* to avoid conflict with cache_control_thread() lock */
	pthread_mutex_lock(&cache_mutex);
	struct cache_item *item = 
		(struct cache_item *)g_hash_table_lookup(cache, remotepath2);
	if (item != NULL) {
		g_hash_table_remove(cache, remotepath2);
		if (debug_mode == true)
			printf("** removed cache item for '%s'\n", remotepath2);
	}
	pthread_mutex_unlock(&cache_mutex);
	FREE(remotepath2);
}


/* looks at the cache for the wanted item. if it's found and not already timed
 * out, the "struct stat *stat" is pointing to the wanted item's stat. 
 * returns 0 on success or -1 on error. */
int cache_get_item(struct stat *stat, const char *remotepath)
{
	int ret = -1;
	assert(remotepath && stat);

	/* generalize the path by removing an ending slash and unescaping it */
	char *remotepath_tmp = remove_ending_slash(remotepath);
	char *remotepath2 = ne_path_unescape(remotepath_tmp);
	FREE(remotepath_tmp);
	if (remotepath2 == NULL) {
		printf("## ne_path_unescape() error in %s()!\n", __func__);
		return -1;
	}

	/* to avoid conflict with cache_control_thread() lock */
	pthread_mutex_lock(&cache_mutex);
	if (g_hash_table_lookup(cache, remotepath2) != NULL) {
		struct cache_item *item =
			(struct cache_item *)g_hash_table_lookup(cache, remotepath2);
		pthread_mutex_unlock(&cache_mutex);
		/* used cached item, if it's not timed out */
		if (!cache_item_timed_out(item->timeout)) {
			*stat = item->stat;
			ret = 0;
			if (debug_mode == true)
				printf("** cache hit for '%s'\n", remotepath2);
		/* if this cache item has timed out, remove it */
		} else {
			if (debug_mode == true)
				printf("** cache item timed out '%s'\n", remotepath2);
			cache_delete_item(remotepath2);
		}
	} else {
		pthread_mutex_unlock(&cache_mutex);
		if (debug_mode == true)
			printf("** <no> cache hit for '%s'\n", remotepath2);
	}
	FREE(remotepath2);
	return ret;
}



src/wdfs-main.h
#ifndef WDFSMAIN_H_
#define WDFSMAIN_H_

#define FUSE_USE_VERSION 25

#include <fuse.h>
#include <ne_basic.h>

/* build the neon version, which is not directly exported by the neon library */
#if defined(NE_FEATURE_TS_SSL)	/* true for neon 0.26+  */
	#define NEON_VERSION 26
#elif defined(NE_FEATURE_SSL)	/* true for neon 0.25+  */
	#define NEON_VERSION 25
#else							/* neon 0.24 is the minimal requirement */
	#define NEON_VERSION 24
#endif
/* 	it's also possible to replace the above with the following: 
	(file configure.ac, after the PKG_CHECK_MODULES call)

	case `pkg-config --modversion neon` in
		0.24*) AC_DEFINE(NEON_VERSION, 24,
				[The minor version number of the neon library]) ;;
		0.25*) AC_DEFINE(NEON_VERSION, 25) ;;
		*)     AC_DEFINE(NEON_VERSION, 26) ;;
	esac
*/

typedef enum {
	true 	= 1,
	false 	= 0
} bool_t;

/* used as mode for unify_path() */
enum {
	ESCAPE,
	UNESCAPE
};

struct wdfs_conf {
	/* the name of the wdfs executable */
	char *program_name;
	/* if set to "true" wdfs specific debug output is generated */
	bool_t debug;
	/* if set to "true" every certificate is accepted without asking the user */
	bool_t accept_certificate;
	/* username of the webdav resource */
	char *username;
	/* password of the webdav resource */
	char *password;
	/* if set to "true" enables http redirect support */
	bool_t redirect;
	/* if set to "true" enables transparent access to all svn revisions in
	 * a repository thru a virtual directory. */
	bool_t svn_mode;
	/* locking mode of files */
	int locking_mode;
	/* timeout for a lock in seconds */
	int locking_timeout;
	/* address of the webdav resource we are connecting to */
	char * webdav_resource;
};

extern struct wdfs_conf wdfs;

/* look at wdfs-main.c for comments on these extern variables */
extern const char *project_name;
extern char *remotepath_basedir;

/* used by wdfs_readdir() and by svn.h/svn.c to add files to requested 
 * directories by using the fuse filler() method. */
struct dir_item {
	void *buf;
	fuse_fill_dir_t filler;
	char *remotepath;
};

char* remove_ending_slash(const char *in);
char* unify_path(const char *in, int mode);
void free_chars(char **arg, ...);

/* Macro to free things: takes an lvalue and sets it to NULL after freeing. */
#define FREE(x) do { if ((x) != NULL) free((x)); (x) = NULL; } while (0)

#endif /*WDFSMAIN_H_*/
